source/elements/oneMKL/source/domains/spblas/operations/spsv.rst

```diff
@@ -0,0 +1,333 @@
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spsv_alg              alg,
+           oneapi::mkl::sparse::spsv_descr_t          spsv_descr,
+           std::size_t                                &temp_buffer_size,
+       );
+
+       void spsv_optimize(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spsv_alg              alg,
+           oneapi::mkl::sparse::spsv_descr_t          spsv_descr,
+           sycl::buffer<std::uint8_t, 1>              workspace,
+       );
+
+       sycl::event spsv_optimize(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spsv_alg              alg,
+           oneapi::mkl::sparse::spsv_descr_t          spsv_descr,
+           void*                                      workspace,
+           const std::vector<sycl::event>             &dependencies = {},
+       );
+
+       sycl::event spsv(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spsv_alg              alg,
+           oneapi::mkl::sparse::spsv_descr_t          spsv_descr,
+           const std::vector<sycl::event>             &dependencies = {},
+       );
+
+   }
+
+.. container:: section
+
+   .. rubric:: Notes
+
+   - ``spsv_buffer_size`` and ``spsv_optimize`` must be called at least once before ``spsv``
+     with the same arguments. ``spsv`` can then be called multiple times.
```

Comment:
I am suggesting that we make all of the `*_optimize` functions mandatory after all. One big advantage I see is that the user does not need to specify the external workspace again during the operation, only once during the `*_optimize` function. This is a difference compared to cuSPARSE/rocSPARSE, in their case the workspace is always a `void*` but in oneMKL it can be a USM pointer or SYCL buffer so adding the workspace to all the operations creates a lot of needless duplication in my opinion.

A couple of concerns I have thought about:

- Can it be a limitation to have a mandatory `*_optimize` step? Current APIs always have a mandatory step before the computation to at least query the workspace size. Adding a mandatory `optimize` step should not block.
- Is it a limitation to not be able to provide the workspace when calling the operation? I suspect most use-cases re-use the same workspace every time. We could consider another function to reset the workspace stored in the operation descriptor if needed.

Reply:
we should discuss this in person.  It is a good question and one we should definitely work out together.

Reply:
As discussed, I kept `*_optimize` functions mandatory and introduced `no_optimize_alg` in https://github.com/oneapi-src/oneAPI-spec/pull/522/commits/652bfe7592a554cd033c894074a20d391d9d7883